### PR TITLE
NODE-1296: State optimizations

### PR DESCRIPTION
**Pull request by disfenoid**
[JIRA: https://wavesplatform.atlassian.net/browse/NODE-1296](https://wavesplatform.atlassian.net/browse/NODE-1296)

### CODE REVIEW

File: src/main/scala/com/wavesplatform/state/package.scala [hunk 1]
```diff
@@ -163,4 +163,15 @@ package object state {
     )
   }
 
+  object Height extends TaggedType[Int]
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state/package.scala]: 👍

File: src/main/scala/com/wavesplatform/state/package.scala [hunk 2]
```diff
@@ -163,4 +163,15 @@ package object state {
...
+  type Height = Height.Type
+
+  object TxNum extends TaggedType[Short]
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state/package.scala]: 👍

File: src/main/scala/com/wavesplatform/state/package.scala [hunk 3]
```diff
@@ -163,4 +163,15 @@ package object state {
...
+  type TxNum = TxNum.Type
+
+  object AddressId extends TaggedType[BigInt]
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state/package.scala]: 👍

File: src/main/scala/com/wavesplatform/state/package.scala [hunk 4]
```diff
@@ -163,4 +163,15 @@ package object state {
...
+  type AddressId = AddressId.Type
+
+  object TransactionId extends TaggedType[ByteStr]
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state/package.scala]: 👍

File: src/main/scala/com/wavesplatform/database/LevelDBWriter.scala [hunk 5]
```diff
@@ -424,19 +453,21 @@ class LevelDBWriter(writableDB: DB, fs: FunctionalitySettings, val maxCacheSize:
...
             val kTxSeqNr = Keys.addressTransactionSeqNr(addressId)
             val txSeqNr  = rw.get(kTxSeqNr)
-            val kTxIds   = Keys.addressTransactionIds(addressId, txSeqNr)
-            for ((_, id) <- rw.get(kTxIds).headOption; (h, _) <- rw.get(Keys.transactionInfo(id)) if h == currentHeight) {
-              rw.delete(kTxIds)
-              rw.put(kTxSeqNr, (txSeqNr - 1).max(0))
-            }
+            val kTxHNSeq = Keys.addressTransactionHN(addressId, txSeqNr)
+            rw.get(kTxHNSeq)
+              .headOption
```
Comment (phearnot) [inline src/main/scala/com/wavesplatform/database/LevelDBWriter.scala]: `headOption` on an `Option`?

File: src/main/scala/com/wavesplatform/database/LevelDBWriter.scala [hunk 6]
```diff
@@ -710,20 +792,72 @@ class LevelDBWriter(writableDB: DB, fs: FunctionalitySettings, val maxCacheSize:
...
   }
 
-  override def loadBlockBytes(height: Int): Option[Array[Byte]] = {
-    readOnly(_.get(Keys.blockBytes(height)))
+  override def loadBlockBytes(h: Int): Option[Array[Byte]] = readOnly { db =>
+    import com.wavesplatform.crypto._
+
+    val height = Height(h)
+
+    val cDataOffset = 1 + 8 + SignatureLength
```
Comment (phearnot) [inline src/main/scala/com/wavesplatform/database/LevelDBWriter.scala]: `cData` must be consensus data, but it took me a while to figure this out :|

File: src/main/scala/com/wavesplatform/database/LevelDBWriter.scala [hunk 7]
```diff
@@ -710,20 +792,72 @@ class LevelDBWriter(writableDB: DB, fs: FunctionalitySettings, val maxCacheSize:
...
 
-  override def loadBlockHeaderAndSize(height: Int): Option[(BlockHeader, Int)] = {
-    readOnly(_.get(Keys.blockHeader(height)))
+  override def loadBlockHeaderAndSize(height: Int): Option[(BlockHeader, Int)] = readOnly { db =>
+    db.get(Keys.blockHeaderAndSizeAt(Height(height)))
   }
 
-  override def loadBlockHeaderAndSize(blockId: ByteStr): Option[(BlockHeader, Int)] = {
-    readOnly(db => db.get(Keys.heightOf(blockId)).flatMap(h => db.get(Keys.blockHeader(h))))
+  override def loadBlockHeaderAndSize(blockId: ByteStr): Option[(BlockHeader, Int)] = readOnly { db =>
```
Comment (phearnot) [inline src/main/scala/com/wavesplatform/database/LevelDBWriter.scala]: There's no point in creating a read-only DB instance when retrieving just one value from.

File: src/main/scala/com/wavesplatform/database/LevelDBWriter.scala [hunk 8]
```diff
@@ -841,4 +992,49 @@ class LevelDBWriter(writableDB: DB, fs: FunctionalitySettings, val maxCacheSize:
...
+    val prefix = ByteBuffer
+      .allocate(6)
+      .putShort(Keys.TransactionInfoPrefix)
+      .putInt(h)
+      .array()
+
+    val iterator = db.iterator
+
+    try {
+      iterator.seek(prefix)
```
Comment (phearnot) [inline src/main/scala/com/wavesplatform/database/LevelDBWriter.scala]: Use `DBExt.iterateOver`

File: src/main/scala/com/wavesplatform/database/Keys.scala [hunk 9]
```diff
@@ -123,4 +101,58 @@ object Keys {
   def changedDataKeys(height: Int, addressId: BigInt): Key[Seq[String]] =
     Key("changed-data-keys", hAddr(49, height, addressId), readStrings, writeStrings)
 
+  // New Keys
```
Comment (phearnot) [inline src/main/scala/com/wavesplatform/database/Keys.scala]: Is there really a reason to skip 51 prefix numbers?

File: waves-mainnet.conf [hunk 10]
```diff
@@ -68,7 +68,7 @@ waves {
     port = 6869
 
     # Hash of API key string
-    api-key-hash = "H6nsiifwYKYEx6YzYD7woP1XCn72RVvx6tC1zjjLXqsu"
+    api-key-hash = "7AshJ286oqdpTYoFbEWXDEqY313qYiaFHMvKYTGXsgwY"
```
Comment (phearnot) [inline waves-mainnet.conf]: Are mainnet config changes really a part of this task?

**Pull request merged**